piet-cairo/src/text.rs

```diff
@@ -210,235 +357,149 @@ impl TextLayout for CairoTextLayout {
-            .unwrap_or_else(|| self.font.extents().height);
-        self.size = Size::new(width, height);
-        self.trailing_ws_width = ws_width;
-
-        Ok(())
-    }
-}
-
-impl fmt::Debug for CairoTextLayout {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.debug_struct("CairoTextLayout").finish()
+        HitTestPosition::new(point, line)
     }
 }
 
-// NOTE this is the same as the old, non-line-aware version of hit_test_point
-// Future: instead of passing Font, should there be some other line-level text layout?
-fn hit_test_line_point(font: &ScaledFont, text: &str, point: Point) -> HitTestPoint {
-    // null case
-    if text.is_empty() {
-        return HitTestPoint::default();
-    }
-
-    // get bounds
-    // TODO handle if string is not null yet count is 0?
-    let end = UnicodeSegmentation::graphemes(text, true).count() - 1;
-    let end_bounds = match get_grapheme_boundaries(font, text, end) {
-        Some(bounds) => bounds,
-        None => return HitTestPoint::default(),
-    };
-
-    let start = 0;
-    let start_bounds = match get_grapheme_boundaries(font, text, start) {
-        Some(bounds) => bounds,
-        None => return HitTestPoint::default(),
-    };
-
-    // first test beyond ends
-    if point.x > end_bounds.trailing {
-        return HitTestPoint::new(text.len(), false);
-    }
-    if point.x <= start_bounds.leading {
-        return HitTestPoint::default();
-    }
-
-    // then test the beginning and end (common cases)
-    if let Some(hit) = point_x_in_grapheme(point.x, &start_bounds) {
-        return hit;
-    }
-    if let Some(hit) = point_x_in_grapheme(point.x, &end_bounds) {
-        return hit;
-    }
+impl CairoTextLayout {
+    fn update_width(&mut self, new_width: impl Into<Option<f64>>) -> Result<(), Error> {
+        if let Some(new_width) = new_width.into() {
+            let pango_width = new_width * pango::SCALE as f64;
+            self.pango_layout.set_width(pango_width as i32);
+        } else {
+            /*
+             * NOTE: -1 is the default value, however `update_width` *could*
+             * be called any number of times with different values so we need
+             * to make sure to reset back to default whenever we get no width
+             */
+            self.pango_layout.set_width(-1);
+        }
 
-    // Now that we know it's not beginning or end, begin binary search.
-    // Iterative style
-    let mut left = start;
-    let mut right = end;
-    loop {
-        // pick halfway point
-        let middle = left + ((right - left) / 2);
-
-        let grapheme_bounds = match get_grapheme_boundaries(font, text, middle) {
-            Some(bounds) => bounds,
-            None => return HitTestPoint::default(),
-        };
+        let mut line_metrics = Vec::new();
+        let mut y_distance = 0.;
+        let mut widest_logical_width = 0;
+        for line_index in 0..self.pango_layout.get_line_count() {
```

Comment:
so instead of doing this manually, I think we should use `PangoLayoutIter` to iterate through lines; this will also let you access the actual baseline values, that are exposed on that type.